File: 1306-jump-game-iii/1306-jump-game-iii.py
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        visited = set([start])
        stack = deque([start])
        while stack:
            idx = stack.pop()
            if arr[idx] == 0: return True
            
            s = idx + arr[idx]
            if s < len(arr) and s not in visited:
                stack.append(s)
                visited.add(s)
                
            d = idx - arr[idx]
            if d >= 0 and d not in visited:
                stack.append(d)
                visited.add(d)
        return False

File: 1306-jump-game-iii/1306-jump-game-iii.cpp
class Solution {
public:
    Solution() {
        ios_base::sync_with_stdio(0);
    }
    
    bool canReach(vector<int>& arr, int start) {
        if (start < 0 || start >= arr.size() || arr[start] < 0) return false;
        arr[start] = -arr[start];
        return !arr[start] || canReach(arr, start + arr[start]) || canReach(arr, start - arr[start]);
    }
};